WIP: Correct completions for quoting and escaping

msftrncs opened the pull request:
# PR Summary

Correct completions (Tab/Intellisense) for quoting and escaping for the various kinds of completions available.

## PR Context

Completers of all types seem to have inconsistent quoting and escaping behaviors that often left the resulting completion still unusable, most specially when the completion contained:
 - potential wildcard patterns
 - escape sequence characters
 - curly quotes
 - variable interpolation patterns
 - specific file system conditions

This PR, through a series of commits, brings unity to the completers, implementing methods that were already available in the code and additional methods, where by each type of completer (variable name, member name, argument) utilizes a common quoting and escaping method.

Fix #10006 - Variable (`$` and `@`) name bracing and escaping and empty scope prefixing
Fix #10239 - Variable completion has no results with input `$:`
Fix #10198 - Member name completion quoting and escaping
Fix #9881 - General argument completion quoting and escaping
Fix #10218 - Argument completion of `ValidateSetAttribute` values quoting and escaping
Fix #7569 - `\\Server\Share With Space` quoting and escaping

This PR is still WIP, as it lacks any changes for testing, and it is unknown if it causes any tests to fail.

To quote @bergmeister, be gentle.  :)  This is my first work in C#.

@rkeithhill, from https://github.com/dahlbyk/posh-git/issues/683#issuecomment-501125585, this includes such a method, `[System.Management.Automation.Language.CodeGeneration]::QuoteArgument()`.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

msftrncs committed:
Create method QuoteMemberName

Create method QuoteMemberName, determines if a supplied member name
requires quoting to be used by an accessor, returning the member name
and escaped if required.

msftrncs committed:
Utilize QuoteMemberName codegenerator

Utilize QuoteMemberName from CodeGenerator class to complete member
names so completions are quoted and escaped as required.
Fixes #10198

msftrncs committed:
improve variable name competion results

Created new private method GenerateVariableCompletionText to handle
correct patterning of `$` and `@` variable completions,  bracing when
required, escaping when braced, and adding leading `:` under certain
circumstances.  Primary improvement is regarding detecting the need to
use braces.

Fixes #10006

msftrncs committed:
allow completion starting with `$:`

Enhances variable completion allowing completions to start with `$:`

msftrncs committed:
remove orphaned s_charactersRequiringQuotes

Its purpose has been removed by prior commits, and future use could be
misleading.

msftrncs committed:
Add argument processing methods

Add methods QuoteArgument and EscapeDoubleQuotedStringContent to assist
generating argument completions that are correctly quoted and escaped.

Includes private method to determine if argument content cannot be used
bareword.

msftrncs committed:
Enhance all argument completers

Enhance all argument completers to use QuoteArgument code generator to
insure consistent quoting and escaping by all argument completion
completers.

Fixes #9881
Fixes #7569

msftrncs committed:
Remove CompletionRequiresQuotes Method

Remove CompletionRequiresQuotes method now orphaned by prior commits.  Method was
unreliable at detecting when an argument completion could not be used
bareword.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4822,6 +4573,42 @@ private static void AddUniqueVariable(HashSet<string> hashedResults, List<Comple
+            // or specifically for an unbraced variable that starts with a `?` (but is more than just a `?`)
+            return braceVariable ?
+                "{" + CodeGeneration.EscapeVariableName((hasProvider || !hasColon ? string.Empty : ":") + value) + "}" :
+                (hasProvider || !(startsWithQM || hasColon) ? string.Empty : ":") + value;
```
Can we simplify the return? It is difficult to read.

msftrncs replied:
That's why there is a set of comment lines right before it.  The expression cannot get 'simpler'.  It can be expanded in to the 4 different return expressions it represents.  while each of the expanded return expression will appear simpler, there will be 4 expressions, and three tests.  

When evaluated closely, there are two main returns, and I think the first ternary makes that easy to see; each of the main returns is on its own line, and then in each one is a unique decisions to insert a ':' in front of the variable name being offered for completion so that the completion will work as expected.

When you add the 'prefix +' to the expression it makes the single return statement that much more efficient:
```C#
            return prefix + (braceVariable ?
                "{" + CodeGeneration.EscapeVariableName((hasProvider || !hasColon ? string.Empty : ":") + value) + "}" :
                (hasProvider || !(startsWithQM || hasColon) ? string.Empty : ":") + value);
```

iSazonov replied:
I meant:
```c#
if (braceVariable)
{
    return …
}
else
{
    return …
}
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+
+        /// <summary>
+        /// Single-quote and escape a member name if it requires quoting, otherwise passing it unmodified.
+        /// For example: QuoteMemberName(userContent)
```
Pleae remove.
```suggestion
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        public static string QuoteMemberName(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
```
```suggestion
            {
                return string.Empty;
            }
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+            bool requiresQuote = !value[0].IsIdentifierStart();
+            if (!requiresQuote)
+            {
+                foreach (char c in value.Substring(1))
```
Please use `for` cycle.

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly,
+        /// handling escaping of wildcard patterns when argument is not already taken literally.
+        /// For example: QuoteArgument(userContent, quoteInUse, isLiteralArgument)
```
```suggestion
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        }
+
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly
```
```suggestion
        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly.
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+
+        /// <summary>
+        /// Quote an argument, if needed, or if specifically requested, escaping characters accordingly
+        /// For example: QuoteArgument(userContent, quoteInUse)
```
```suggestion
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        public static string QuoteArgument(string value, char quoteInUse)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
```
```suggestion
          {
                return string.Empty;
          }
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+            if (quoteToUse == 0)
+                if (ShouldArgumentNotBeBareword(value))
+                    // argument value not compatible with bareword
+                    quoteToUse = '\'';
```
Please use follow pattern:
```c#
if
{
    ...
}
else
{
    ...
}
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        private static bool ShouldArgumentNotBeBareword(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                return false;
```
```suggestion
           {
                return false;
           }
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+            //   characters that cannot be in the first position: '@','#','<','>'
+            //   Pattern that cannot be in the first position: /[1-6]>/
+            bool requiresQuote = "@#<>".Contains(value[0]) ||
+                value.Length > 1 && '1' <= value[0] && value[0] <= '6' && value[1] == '>';
```
Please use the pattern:
(uint)(value[0] - '1') <= '6' - '1'

Why do we need last "value[1] == '>'"? It is second position. Or previous comment is wrong?

msftrncs replied:
> Please use the pattern:
> (uint)(value[0] - '1') <= '6' - '1'
Why `unit`?  Does subtraction of [char]'s automatically become unit?

msftrncs replied:
I confirmed that it appears that compiler must upscale during arithmetic, so even though it looks like an upscale, its actually leaving it at the scale it performed it at.

Also, it looks like compiler does not pre-compute literal arithmetic (at least not in debug), so using `5` in place of `'6' - '1'` is another >10% faster yet.

iSazonov replied:
> at least not in debug

You should check release code to get all optimizations. If after that you still a bad code you should report in CoreClr repository because the pattern is used in CoreFX a lot.

> Maybe comment should read

Your comments is still not correct. 

>             //   characters that cannot be in the first position: '@','#','<','>' 
            //   Pattern that cannot be in the first position: /[1-6]>/

It seems it should be:
```
            //   characters that cannot be in the first position: '@','#','<','>' 
            //   Pattern that cannot be in the second position: /[1-6]>/
```

msftrncs replied:
Note that this code is outdated and was changed greatly, both the code and the comment referenced above had to change because of more patterns that needed to be checked.

I think there may still be some confusion, the pattern `/[1-6]>/` is two characters long, and it must NOT start in the first position of the 'value', (or value[0] and value[1] respectively).

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+            if (!requiresQuote)
+            {
+                bool lastCharWasDollar = false;
+                foreach (char c in value)
```
Please use `for` cycle.

msftrncs replied:
Explain `for cycle`?

I built the loops referencing the example in the same file, `EscapeSingleQuotedStringContent()`

iSazonov replied:
No need to use an enumarator to access chars from a string. 
I means value[i].

msftrncs replied:
so a for loop of

```C#
// skipping first char (special case)
for( int i = 1; i < value.Length; i++ )
{
    value[i]  ...
}
```

will be faster than the enumerator?  Is this something that can only be performance tested in release?

iSazonov replied:
> will be faster than the enumerator? 

You could test with BenchmarkDotNet. 
I say only that we have no need to allocate an enumerator in the simple code.

msftrncs replied:
I finally got to do some benchmarks.  It does appear that the method 'indexing' is probably both the fastest and the most flexible, particularly when the strings are short.  Once the strings reach 10000's of characters, 'indexing' and 'foreach' methods differences become negligible, though system background tasks had a larger effect at skewing the results between runs.

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+        public static string EscapeDoubleQuotedStringContent(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+                return string.Empty;
```
```suggestion
           {
                return string.Empty;
           }
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+                if (lastCharWasDollar)
+                {
+                    if (c.IsVariableStart() || c == '{' || c == '(')
+                        sb.Append('`');
```
```suggestion
                   {
                        sb.Append('`');
                   }
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+
+            StringBuilder sb = new StringBuilder(value.Length);
+            bool lastCharWasDollar = false;
+            foreach (char c in value)
```
Please use `for` cycle.

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+                        sb.Append('`');
+                    sb.Append('$');
+                }
+                lastCharWasDollar = c == '$';
```
```suggestion

                lastCharWasDollar = c == '$';
```

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+                    sb.Append(c);
+                    if (c.IsDoubleQuote() || c == '`')
+                        // double-up quotes & backticks to escape them
+                        sb.Append(c);
```
Please use { }.

iSazonov commented on src/System.Management.Automation/engine/lang/codegen.cs:
```diff
@@ -106,5 +106,154 @@ public static string EscapeVariableName(string value)
+                }
+            }
+            if (lastCharWasDollar)
+                sb.Append('$');
```
```suggestion
            {
                sb.Append('$');
            }
```

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -228,30 +228,12 @@ internal static CompletionResult GetCommandNameCompletionResult(string name, obj
 
             syntax = string.IsNullOrEmpty(syntax) ? name : syntax;
-            bool needAmpersand;
+            name = CodeGeneration.QuoteArgument(name, quote == string.Empty ? (char)0 : quote[0], false);
```
Would it be acceptable to rewrite `HandleSingleAndDoubleQuote()`:

https://github.com/PowerShell/PowerShell/blob/8f5b2c241ec1bc0e9c97176b591cf342dcc10da2/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L6489-L6519

with this:

```C#
        internal static char HandleDoubleAndSingleQuote(ref string wordToComplete)
        {
            char quote = (char)0;

            if (!string.IsNullOrEmpty(wordToComplete) && (wordToComplete[0].IsSingleQuote() || wordToComplete[0].IsDoubleQuote()))
            {
                quote = wordToComplete[0].IsSingleQuote() ? '\'' : '"';
                int length = wordToComplete.Length;

                wordToComplete = length == 1 ? string.Empty :
                    (quote.IsSingleQuote() ? wordToComplete[length - 1].IsSingleQuote() : wordToComplete[length - 1].IsDoubleQuote()) ?
                        wordToComplete.Substring(1, length - 2) : wordToComplete.Substring(1);

            }

            return quote;
        }
```
and then eliminate the kludge?  This would change the entire code over to using `char` to pass the quote.

iSazonov replied:
You check quote == 0 only once in "QuoteArgument". I guess you could use quote.Empty and don't use the zero value.

msftrncs replied:
It comes down to, do you use a string to store just three values, empty, `'` or `"`. or is there a better way?  Using a string means you have to use [0] references to get at the character (after you've confirmed that the string isn't empty) for using `IsSingleQuote` and `IsDoubleQuote`.

In a sense, its a glorified enum.  Is(Single/Double)Quote's are overkill because at this point it can only contain 3 values, not any of the other 7 quotes (at least as defined by HandleSingleAndDoubleQuote).

msftrncs commented:
Is there a lint I am supposed to be using?   I don't think any of the CodeFactor / Codacy issues would be there if there if I had the linting warnings.  I didn't find any references to a lint in the contributing guide.

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4685,7 +4452,7 @@ internal static List<CompletionResult> CompleteVariable(CompletionContext contex
             {
                 provider = wordToComplete.Substring(0, colon + 1);
-                if (s_variableScopes.Contains(provider, StringComparer.OrdinalIgnoreCase))
+                if (colon == 0 || s_variableScopes.Contains(provider, StringComparer.OrdinalIgnoreCase))
```
per https://github.com/PowerShell/PowerShell/pull/10226/commits/033ef6cd6bf0b3f8a78a9b618f98334cd5aa7626:

```
allow completion starting with `$:` … 
Enhances variable completion allowing completions to start with `$:`
```

msftrncs replied:
Added issue #10239.

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4822,6 +4573,42 @@ private static void AddUniqueVariable(HashSet<string> hashedResults, List<Comple
+
+            bool startsWithQM = value[0] == '?' && value.Length > 1;
+            bool lastCharWasColon = false, hasColon = false;
+            bool braceVariable = !value[0].IsVariableStart() || value.Length != 1 && (value[0] == '$' || value[0] == '^' || (hasProvider && startsWithQM));
```
> Should IsVariableStart() be on first place for performance reasons?

Because its more likely that the first character will be < 128 and will usually be valid and unlikely to be '$', '^' or '?' even though most names are more than 1 character, it would seem to me that keeping IsVariableStart() at the start of the expression is best, providing the shortest path on the most number of iterations.  If there was some minor tweaking it could be to make startsWithQM first in the last group, instead of hasProvider, and then make that group first in its group.

The expression is at its simplest.

result is TRUE if
- value[0] Is NOT Variable Starting character OR ELSE
- value.Length > 1 AND ALSO
  - value[0] is '$' OR ELSE
  - value[0] is '^'  OR ELSE
      - hasProvider AND ALSO
      - startsWithQM

I could put some help in front of this method to explain what it is doing, would possibly help with understanding the return value as well.

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4822,6 +4573,42 @@ private static void AddUniqueVariable(HashSet<string> hashedResults, List<Comple
+            {
+                lastCharWasColon = !hasProvider && value[0] == ':';
+                hasColon = lastCharWasColon;
+                foreach (char c in value.Substring(1))
```
You mean something like this?

```C#
                // Enumerate through the string, but the first character is already evaluated with special rules
                // so we'll skip it.
                CharEnumerator ce_value = value.GetEnumerator();
                for (ce_value.MoveNext(); ce_value.MoveNext();)
                {
                    char c = ce_value.Current;
```

iSazonov replied:
We could directly access elements in the string without creating enumerator like value[i].

msftrncs committed:
style, fix comments, minor code suggestions

msftrncs committed:
move variable prefix concat into GenerateVariableCompletionText

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4822,6 +4573,42 @@ private static void AddUniqueVariable(HashSet<string> hashedResults, List<Comple
+            bool braceVariable = !value[0].IsVariableStart() || value.Length != 1 && (value[0] == '$' || value[0] == '^' || (hasProvider && startsWithQM));
+            if (!braceVariable)
+            {
+                lastCharWasColon = !hasProvider && value[0] == ':';
```
I just realized a major flaw in my thought process regarding `hasColon` / `lastCharWasColon`, as I was trying to perform two tests, a check for double colons, and regardless of braced or not, if there was any colons in the entire name.  This loop is not set up for that.   Since I need to scan the entire name for a colon, it appears its most efficient to use `Contains(char)`, than it is to make the current loop continue scanning for the colon even if it has found the brace condition.  Unfortunately it still needs to look for `::` as one of the possible brace conditions,

msftrncs commented:
Noting, this PR may conflict with #7407, as this PR attempts to do some of the same things, but not exclusively to file paths.

msftrncs committed:
Correct WildCard Escape Escape Char

msftrncs committed:
Include quoting arg starting with dashes

Include in ShouldArgumentNotBeBareword, returning `true` when argument
would appear to be a parameter name, which will cause such arguments
values to require quoting or escaping.

msftrncs committed:
use enumerator directly in QuoteMemberName

msftrncs committed:
correct and optimize GenerateVariableCompletionText

GenerateVariableCompletionText failed to scan entire variable name for
`:`, if a character requiring a brace was found first.
Optimized character look to skip first character without using substring.

msftrncs committed:
use string for QuoteToUse

Revision to base the passed parameters for the current quote character
to be type string instead of char, the way the previous code worked.

msftrncs committed:
Optimize HandleDoubleAndSingleQuote

Optimize HandleDoubleAndSingleQuote() method, reducing redundant and
unneccessary operations.

msftrncs committed:
correct short-circuit logic with parenthesis

corrected where completions of commands for an argument would incorrectly
include an `&`.

msftrncs commented:
## Broken Tests

Two test patterns are currently broke.  The reason for this is that this PR implements a more specific `$` variable interpolation detection to require the argument to be quoted.  The implementation is that, if a `$` character is not followed by a legitimate variable character, that `$` will not itself require the argument to be quoted.

Patterns failing:
- a file path completion test where `$` appears at the end of the completion
- a Windows only miscellaneous completion test where `$` appears followed by a `\` in the completion.

Both test patterns involve command completions, so the quoting of the path resulted in the completion using the invoke operator `&`.  With this PR's implementation, the quoting and the invoke operator no longer occur.

## Tests Needed
- Member completion quoting tests, these should probably be similar to the two examples that were previously commented in the code, as I haven't found any tests for them yet, in addition to tests relating to issue #10198.
- Variable completion tests
   - test completions of variable names that require braces and quoting and escaping both for access and for use as an argument.
   - test completions of variable names that have colons in their names, both with and without a scope specified.
   - test completions of variable names that start with `?` but with more than one character, with scope, with complex character requring braces, with scope and complex character, without scope and without complex character.
- Argument value completion tests
   - test escaping of wildcard patterns in a completion for a parameter which accepts them. __\*\* IMPORTANT for other work being done around the WildCardPattern class  \*\*__
   - test quoting and escaping of curly quoting characters.
   - On Unix OS's test completion of a file name with wildcard characters `*` and `?` (with a parameter supporting wildcards) and doublequotes.  Tests should also be added to other areas to insure these files or folders are accessible, but that is not the scope of this PR (these tests will need to create said files though).
 - test `ValidateSetAttribute` completed arguments that require quoting, especially regarding those that could be confused as a parameter name when not quoted.  Include variants related to issue #10218.

We'll keep adding to this list to track the tests fixed, needed and added.

iSazonov commented:
> Is there a lint I am supposed to be using?

CodeFactor is based on StyleCop - you can run it locally. Usually we don't request to fix style issue which is not in your code. If there is a lot of style issues please pull new PR to ifx them - we don't want to mess fuctional and style changes in one PR.

iSazonov commented:
@kwkam Could you please review the PR too?

msftrncs commented:
> CodeFactor is based on StyleCop - you can run it locally. Usually we don't request to fix style issue which is not in your code.

I did notice that a lot of the style errors were not from my code changes (at first I thought they were all in my code, probably because it showed new issues first), and that would make it difficult to let a format-document command run, but it would sure help if I could format-selection and get the easy ones taken care of, and then a lint would show the other suggestions.  That's how the TS/JS repositories I have contributed to work, and its all in VS Code.

I was able to turn on the 'editor settings' feature of the C# extension and it did change some of the formatting, but it still doesn't warn me that this repository would prefer all IF statements braced, comments spaced out, statements after IF statements spaced out, etc...

SSJ9899TDylan commented:
Holy shit!

On Mon, 29 Jul 2019 at 4:51 pm, Carl Morris <notifications@github.com>
wrote:

> *@msftrncs* commented on this pull request.
> ------------------------------
>
> In src/System.Management.Automation/engine/lang/codegen.cs
> <https://github.com/PowerShell/PowerShell/pull/10226#discussion_r308076392>
> :
>
> > +            // quote and escape argument as needed
> +            return quoteToUse + (quoteToUse.IsDoubleQuote() ? EscapeDoubleQuotedStringContent(value) :
> +                EscapeSingleQuotedStringContent(value)) + quoteToUse;
> +        }
> +
> +        // brute force method to determine when an argument's value cannot be used bareword
> +        private static bool ShouldArgumentNotBeBareword(string value)
> +        {
> +            if (string.IsNullOrEmpty(value))
> +                return false;
> +
> +            // Rules for bareword arguments:
> +            //   characters that cannot be in the first position: '@','#','<','>'
> +            //   Pattern that cannot be in the first position: /[1-6]>/
> +            bool requiresQuote = "@#<>".Contains(value[0]) ||
> +                value.Length > 1 && '1' <= value[0] && value[0] <= '6' && value[1] == '>';
>
> Note that this code is outdated and was changed greatly, both the code and
> the comment referenced above had to change because of more patterns that
> needed to be checked.
>
> I think there may still be some confusion, the pattern /[1-6]>/ is two
> characters long, and it must NOT start in the first position of the
> 'value', (or value[0] and value[1] respectively).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PowerShell/PowerShell/pull/10226?email_source=notifications&email_token=AIPXOJT3D3G7Q6BN4SM4TF3QB2HODA5CNFSM4IGYD4D2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7ZSOKI#discussion_r308076392>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIPXOJXNW3WUXBF2YBUF3C3QB2HODANCNFSM4IGYD4DQ>
> .
>
-- 
Sent by Dylan Taylor

msftrncs commented on src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:
```diff
@@ -4722,10 +4490,8 @@ internal static List<CompletionResult> CompleteVariable(CompletionContext contex
-                                                : prefix + "{" + provider + name + "}";
-                        AddUniqueVariable(hashedResults, results, completedName, name, tooltip);
+                        AddUniqueVariable(hashedResults, results,
+                            GenerateVariableCompletionText(provider + name, prefix, colon != -1), name, tooltip);
```
I'm considering putting the call to GenerateVariableCompletionText inside AddUniqueVariable, because if the variable is not unique, there is no reason to determine if it needs braced.

msftrncs commented:
I think I need to have an exception in the argument quoting/bareword checker for commands, as they have a slightly different bareword syntax and I am not account for that.

msftrncs committed:
change of foreach loops to for / index loops

msftrncs commented:
That's weird, the Windows test must be skipping the completion tests, because its not failing them, where as the MacOS and Linux tests are failing completion tests (which are currently expected at this time).

iSazonov commented:
@msftrncs It is Newline difference.

iSazonov commented:
@msftrncs Have you plans to continue?

msftrncs commented:
I do, having trouble dedicating time.  I do have new commits for improved use of the parser to determine when quoting is required, specially for command name values.

iSazonov commented:
I see 9-11 tests fails. Is it expected?

msftrncs commented:
> I see 9-11 tests fails. Is it expected?

Yes and no, I expected the '$' tests to fail, because an ending `$` doesn't need to be quoted, but the ending ``` ` ``` should have triggered a requirement to quote.  Those tests previously had not failed.

I also have not yet handled testing, as I am not sure how much of these changes can be specifically tested.

msftrncs committed:
Implement new xRequiresQuote in CodeGen

Implement a xRequiresQuote in CodeGen, similar to the previous
CompletionRequiresQuote, but in 3 parts, and with much more thorough
checks for parsing conditions that indicate the value needs to be quoted.

msftrncs commented:
At this time, I do expect the completion tests that involve a `$` to fail, as the parser doesn't see them as separate tokens so no need to quote.  However, I need to understand the result regarding the `configuration` keyword.   It appears to be 1 of 3 keywords receiving special treatment.

3052 commented:
FYI #9881 this seems to be fixed as of version 7.4.0

iSazonov commented:
Sorry, I'm closing it, because it's a long time ago. If there is something that has not been fixed yet in recent version, it is better to open a new PR. Thanks!
